text/0000-offline-node-improved-trajectories-bags.md

```diff
@@ -0,0 +1,106 @@
+| __B.__ Multiple identical robots  (using the same sensor topics/frames) | __2.__ Multiple trajectories, necessarily sequential | __b.__ Multiple continuous bags  (like a single bag) - a bag aggregate |
+| __C.__ Multiple different robots  (different sensor topics/frames)      | __3.__ Multiple trajectories,  possibly concurrent   | __c.__ Multiple bags                                                   |
+
+The online node currently supports the following combinations:
+__A1__, __A/B 2__ and __C 2/3__ (using `StartTrajectory` and `FinalizeTrajectory` service calls)
+
+The offline node currently supports: __A1a__, __A/B 2c__.
+
+The online node can never support __B3__, because only single robot's data can be published on a topic at a time.
+Also, __B3__ and __a/b__ is impossible for the same reasons.
+
+Note that some combinations are invalid, such as __A3__ (a single robot can't be on multiple concurrent trajectories), and __B/C 1__ (there can't be a single trajectory if there are multiple robots).
+Furthermore, __A/B 2 a/b__ would require us to be able to specify when a trajectory stops and begins within the same bag aggregate and on the same sensor data topics, which would be too complicated and out of scope.
+
+The aim of the RFC is to enable the user to use the offline node for __C2/3__, like the online node; to preserve the ability of doing __A/B 2c__; and, to possibly offer additional functionality. For example, __b__ (bag aggregates), and __B3c__, which would otherwise be impossible with the online node.
+
+## Approach
+[approach]: #approach
+
+The proposed offline node configuration for multiple, possibly concurrent trajectories (__3__) should closely follow the way of the online node for achieving this, listed here for reference:
+  - Performing multiple calls of the `StartTrajectory` service, with possibly different `TrajectoryOptions` and necessarily different sensor data topics (__C__); all is specified in the service request.
+  - Alternatively, the `start_trajectory` helper node is invoked multiple times, with possibly different `.lua` files for `TrajectoryOptions` and necessarily different sensor data topics (__C__); the recently merged PR googlecartographer/cartographer_ros#584 enables specifying different topics via topic remapping.
+  - Specially, if a trajectory is finalized, we may start another independent trajectory which uses the same sensor data topics and possibly different `TrajectoryOptions` (__B2__).
+  This is already achievable in the offline node by simply loading multiple bags, which creates a trajectory for each bag, although `TrajectoryOptions` are necessarily the same for all trajectories (__B2c__). 
+  The ability to do this has to be preserved or expanded.
+
+If we suppose that there is a mandated 1:1 relationship between bags and trajectories in the offline node, the following configuration can be proposed.
+
+### Configuration #1
```

Comment:
I would advise to keep the design and scope limited to what is absolutely necessary to solve your usecase. Sure it would be nice to have the additional functionality of options 2 and 3, but with some bag surgery it seems you can reduce both of these to option 1. I would suggest to move ahead with that.

Reply:
+1. All functionality needs to be taught somehow. Cartographer is already a complex system, so adding more features will make it even harder to understand. Multi trajectory is not a self evident concept in general, so making it as simple to understand as possible is important.

My suggestion: Move this explanation to discussion points, explain why handling one trajectory per bag is sufficient with some bag filtering and that we limited the scope to this, then simplify the rest of the rfc to only deal with this use case.